File: companion/android/app/src/main/kotlin/io/allonsy/kokoro/programme/Programme.kt
package io.allonsy.kokoro.programme

enum class Quand(val cle: String) {
    AUJOURDHUI("aujourdhui"),
    AU_BESOIN("au_besoin"),
    SANS_DATE("sans_date"),
}

enum class Rubrique(val cle: String) {
    CRISE("crise"),
    THERAPIE("therapie"),
    BILAN("bilan"),
    DOCUMENTATION("documentation"),
}

// Qui tient le téléphone. Sur une carte portée par l'aidant, Kokoro tient la cadence et ne lui demande jamais de juger.
enum class Porteur(val cle: String) {
    PATIENT("patient"),
    AIDANT("aidant"),
}

// Guidée : une étape à la fois, dans l'ordre, jusqu'à l'issue. Libre : un sommaire, chaque étape se ferme et y revient.
enum class Allure { GUIDEE, LIBRE }

enum class Unite(val cle: String) {
    BRUTE("brute"),
    MINUTES("minutes"),
    HEURES("heures"),
    KILOS("kilos"),
}

data class Choix(val valeur: Double, val libelle: String)

data class Compteur(
    val depart: Double,
    val pas: Double,
    val grandPas: Double,
    val minimum: Double,
    val unite: Unite,
)

sealed interface Saisie {
    data class Fermee(val choix: List<Choix>) : Saisie

    data class Reglee(val compteur: Compteur) : Saisie
}

sealed interface Etape {
    data class Info(val texte: String, val montrable: Boolean = false) : Etape

    data class Question(
        val id: String,
        val enonce: String,
        val precision: String?,
        val saisie: Saisie,
        // Repart de la dernière valeur donnée à cette question, jamais d'un ressenti à retrouver.
        val reprise: Boolean = false,
    ) : Etape

    data class Note(val id: String, val enonce: String, val precision: String?) : Etape

    data class Minuteur(val secondes: Int, val consigne: String?, val pour: Porteur?) : Etape

    data class Checklist(val enonce: String, val lignes: List<String>) : Etape

    data class Confirmation(val libelle: String) : Etape
}

// Une étape qui n'écrit rien ne fait pas rendre la carte : une fiche lue ne renvoie pas plus qu'avant.
val Etape.rend: Boolean
    get() = this is Etape.Question || this is Etape.Note || this is Etape.Minuteur || this is Etape.Confirmation

data class Reperes(
    val id: String,
    val titre: String,
    val rubrique: Rubrique,
    // Nul sur une carte du dossier Bilan, et sur elle seule : sa date appartient au document, pas à l'assiduité de Xavier.
    val quand: Quand?,
    val dureeMinutes: Int?,
)

sealed interface Carte {
    val reperes: Reperes

    data class Panneau(
        override val reperes: Reperes,
        val etapes: List<Etape>,
        val porteur: Porteur = Porteur.PATIENT,
        val allure: Allure = Allure.GUIDEE,
        val signalArret: String? = null,
        val arret: List<String> = emptyList(),
    ) : Carte

    // date non nulle : le document vit dans bilans/, pas dans la bibliothèque — canal distinct, contrôles distincts.
    data class Pdf(override val reperes: Reperes, val document: String, val date: String? = null) : Carte
}

val Carte.id: String get() = reperes.id

val Carte.titre: String get() = reperes.titre

val Carte.rubrique: Rubrique get() = reperes.rubrique

val Carte.quand: Quand? get() = reperes.quand

val Carte.dureeMinutes: Int? get() = reperes.dureeMinutes

data class Programme(val version: Int, val cartes: List<Carte>)

val PROGRAMME_ABSENT = Programme(version = 0, cartes = emptyList())

private val KEBAB = Regex("""^[a-z0-9-]+$""")

private val DATE = Regex("""^\d{4}-\d{2}-\d{2}$""")

private val DERNIER_CRITERE = Regex("""ne sais pas quoi faire""", RegexOption.IGNORE_CASE)

fun lireProgramme(json: String): Programme {
    val racine = lireJson(json) ?: return PROGRAMME_ABSENT

    return Programme(
        version = racine.entier("version") ?: 0,
        cartes = racine.elements("cartes").mapNotNull(::carte),
    )
}

fun Programme.cartesDe(rubrique: Rubrique): List<Carte> =
    cartes.filter { it.rubrique == rubrique && it !is Carte.Pdf }

// Un PDF vit sur Documentation quelle que soit sa rubrique ; un bilan a son propre écran.
fun Programme.documents(): List<Carte.Pdf> =
    cartes.filterIsInstance<Carte.Pdf>().filter { it.date == null }

fun Programme.bilans(): List<Carte.Pdf> =
    cartes.filterIsInstance<Carte.Pdf>().filter { it.date != null }.sortedByDescending { it.date }

fun moisDe(date: String): String = date.take(7)

private fun carte(valeur: Valeur): Carte? {
    val reperes = reperes(valeur) ?: return null
    val bilan = reperes.rubrique == Rubrique.BILAN

    return when (valeur.texte("type")) {
        "pdf" -> pdf(reperes, valeur, bilan)
        "panneau" -> if (bilan) null else panneau(reperes, valeur)
        else -> null
    }
}

private fun reperes(valeur: Valeur): Reperes? {
    val id = valeur.texte("id")?.takeIf { it.matches(KEBAB) } ?: return null
    val titre = valeur.permis("titre") ?: return null
    val rubrique = Rubrique.entries.firstOrNull { it.cle == valeur.texte("rubrique") } ?: return null
    val quand = Quand.entries.firstOrNull { it.cle == valeur.texte("quand") }

    return Reperes(
        id = id,
        titre = titre,
        rubrique = rubrique,
        quand = quand,
        dureeMinutes = valeur.entier("duree_minutes"),
    )
}

// Une carte rangée au Bilan sans date n'aurait pas de place à l'écran, et disparaîtrait en silence.
private fun pdf(reperes: Reperes, valeur: Valeur, bilan: Boolean): Carte.Pdf? {
    val document = valeur.texte("document")?.takeIf { it.matches(KEBAB) } ?: return null
    val date = valeur.texte("date")?.takeIf { it.matches(DATE) }

    if (bilan != (date != null)) return null
    if (bilan != (reperes.quand == null)) return null

    return Carte.Pdf(reperes = reperes, document = document, date = date)
}

private fun panneau(reperes: Reperes, valeur: Valeur): Carte.Panneau? {
    if (reperes.quand == null) return null
    if (valeur.booleen("sortie_libre") != true) return null

    val porteur = Porteur.entries.firstOrNull { it.cle == valeur.texte("porteur") } ?: Porteur.PATIENT
    val lues = valeur.elements("etapes").map { etape(it, porteur) }
    if (lues.isEmpty() || lues.any { it == null }) return null

    val carte = Carte.Panneau(
        reperes = reperes,
        etapes = lues.filterNotNull(),
        porteur = porteur,
        signalArret = valeur.permis("signal_arret"),
        arret = textesPermis(valeur, "arret").orEmpty(),
    )

    return if (porteur == Porteur.AIDANT && !tenable(carte)) null else carte
}

// L'aidant ne peut ni corriger ni improviser : une carte qu'elle tient tombe entière s'il lui manque de quoi s'arrêter.
private fun tenable(carte: Carte.Panneau): Boolean {
    val minuteurs = carte.etapes.filterIsInstance<Etape.Minuteur>()

    return carte.signalArret != null &&
        carte.arret.size >= 2 &&
        DERNIER_CRITERE.containsMatchIn(carte.arret.last()) &&
        carte.etapes.first() is Etape.Checklist &&
        minuteurs.isNotEmpty() &&
        minuteurs.all { it.pour != null && it.consigne != null }
}

private fun etape(valeur: Valeur, porteur: Porteur): Etape? = when (valeur.texte("type")) {
    "info" -> valeur.permis("texte")?.let {
        Etape.Info(texte = it, montrable = valeur.booleen("montrable") == true)
    }

    "question" -> question(valeur)
    "note" -> note(valeur)
    "minuteur" -> minuteur(valeur, porteur)
    "checklist" -> checklist(valeur)
    "confirmation" -> valeur.permis("libelle")?.let(Etape::Confirmation)
    else -> null
}

// Un item perdu produit un score faux, donc faussement rassurant : une question amputée emporte la carte entière.
private fun question(valeur: Valeur): Etape.Question? {
    val id = valeur.texte("id")?.takeIf { it.matches(KEBAB) } ?: return null
    val enonce = valeur.permis("enonce") ?: return null
    val precision = valeur.permisSiPresent("precision") ?: return null
    val saisie = saisie(valeur) ?: return null

    return Etape.Question(
        id = id,
        enonce = enonce,
        precision = precision.ifEmpty { null },
        saisie = saisie,
        reprise = valeur.booleen("reprise") == true,
    )
}

private fun saisie(valeur: Valeur): Saisie? {
    val compteur = valeur.champ("compteur")
    val lus = valeur.elements("choix").map(::choix)

    if (compteur != null && lus.isNotEmpty()) return null
    if (compteur != null) return compteur(compteur)?.let(Saisie::Reglee)
    if (lus.size < 2 || lus.any { it == null }) return null

    return Saisie.Fermee(lus.filterNotNull())
}

private fun compteur(valeur: Valeur): Compteur? {
    val pas = valeur.nombre("pas")?.takeIf { it > 0.0 } ?: return null
    val grandPas = valeur.nombre("grand_pas")?.takeIf { it >= pas } ?: return null
    val minimum = valeur.nombre("minimum") ?: return null
    val depart = valeur.nombre("depart")?.takeIf { it >= minimum } ?: return null
    val unite = Unite.entries.firstOrNull { it.cle == valeur.texte("unite") } ?: return null

    return Compteur(depart = depart, pas = pas, grandPas = grandPas, minimum = minimum, unite = unite)
}

private fun choix(valeur: Valeur): Choix? {
    val chiffre = valeur.nombre("valeur") ?: return null
    val libelle = valeur.permis("libelle") ?: return null

    return Choix(valeur = chiffre, libelle = libelle)
}

private fun note(valeur: Valeur): Etape.Note? {
    val id = valeur.texte("id")?.takeIf { it.matches(KEBAB) } ?: return null
    val enonce = valeur.permis("enonce") ?: return null
    val precision = valeur.permisSiPresent("precision") ?: return null

    return Etape.Note(id = id, enonce = enonce, precision = precision.ifEmpty { null })
}

private fun minuteur(valeur: Valeur, porteur: Porteur): Etape.Minuteur? {
    val secondes = valeur.entier("secondes")?.takeIf { it > 0 } ?: return null
    val consigne = valeur.permisSiPresent("consigne") ?: return null
    val pour = Porteur.entries.firstOrNull { it.cle == valeur.texte("pour") }

    if (porteur == Porteur.AIDANT && pour == null) return null

    return Etape.Minuteur(secondes = secondes, consigne = consigne.ifEmpty { null }, pour = pour)
}

private fun checklist(valeur: Valeur): Etape.Checklist? {
    val enonce = valeur.permis("enonce") ?: return null
    val lignes = textesPermis(valeur, "lignes")?.takeIf { it.isNotEmpty() } ?: return null

    return Etape.Checklist(enonce = enonce, lignes = lignes)
}

private fun textesPermis(valeur: Valeur, cle: String): List<String>? {
    val lus = valeur.elements(cle).map { element ->
        (element as? Valeur.Texte)?.contenu?.takeIf { it.isNotBlank() && estPermis(it) }
    }

    return if (lus.any { it == null }) null else lus.filterNotNull()
}

private fun Valeur.permis(cle: String): String? = texte(cle)?.takeIf { it.isNotBlank() && estPermis(it) }

// Distingue « champ absent » (permis, rendu vide) de « champ fautif » (null, qui écarte la carte).
private fun Valeur.permisSiPresent(cle: String): String? = if (champ(cle) == null) "" else permis(cle)

File: psy/scripts/schemas/programme.ts
import { z } from 'zod';

export const TYPES = ['panneau', 'pdf'] as const;

export const ETAPES = ['info', 'question', 'note', 'minuteur', 'checklist', 'confirmation'] as const;

// Une carte qui porte au moins une de ces etapes rend une reponse : c'est elle qui fait agir.
export const ETAPES_QUI_RENDENT = ['question', 'note', 'minuteur', 'confirmation'] as const;

const KEBAB = /^[a-z0-9-]+$/;
const JOUR = /^\d{4}-\d{2}-\d{2}$/;
const PHQ9 = /phq[\s-]?9/i;
const DERNIER_CRITERE = /ne sais pas quoi faire/i;

const identifiant = (precision = '') =>
  z.string({ error: `champ absent${precision}` }).regex(KEBAB, { error: `hors kebab-case${precision}` });

const texteNonVide = (precision = '') =>
  z.string({ error: `champ absent${precision}` }).regex(/\S/, { error: `champ vide${precision}` });

const jour = (precision = '') =>
  z.string({ error: `champ absent${precision}` }).regex(JOUR, { error: `hors AAAA-MM-JJ${precision}` });

const commun = {
  id: identifiant(),
  titre: texteNonVide(),
  duree_minutes: z.number({ error: "duree_minutes n'est pas un nombre" }).optional(),
};

const quand = z.enum(['aujourdhui', 'au_besoin', 'sans_date'], {
  error: 'quand attendu : aujourdhui, au_besoin ou sans_date',
});

const rubrique = z.enum(['crise', 'therapie', 'documentation'], {
  error:
    "rubrique attendue : crise, therapie ou documentation — la rubrique bilan est reservee aux cartes du dossier Bilan, rangee la une autre carte n'aurait pas de place a l'ecran",
});

const porteur = z.enum(['patient', 'aidant'], { error: 'porteur attendu : patient ou aidant' });

const info = z.object(
  {
    type: z.literal('info'),
    texte: texteNonVide(),
    montrable: z.boolean({ error: "montrable n'est pas un booleen" }).optional(),
  },
  { error: "ce n'est pas un objet" },
);

const choix = z.object(
  {
    valeur: z.number({ error: '« valeur » absente ou non numerique' }),
    libelle: texteNonVide(),
  },
  { error: "ce n'est pas un objet" },
);

const compteur = z
  .object(
    {
      depart: z.number({ error: '« depart » absent ou non numerique' }),
      pas: z.number({ error: '« pas » absent ou non numerique' }).positive({ error: '« pas » doit etre positif' }),
      grand_pas: z.number({ error: '« grand_pas » absent ou non numerique' }),
      minimum: z.number({ error: '« minimum » absent ou non numerique' }),
      unite: z.enum(['brute', 'minutes', 'heures', 'kilos'], {
        error: 'unite attendue : brute, minutes, heures ou kilos',
      }),
    },
    { error: "ce n'est pas un objet" },
  )
  .superRefine((regle, ctx) => {
    if (regle.grand_pas < regle.pas) {
      ctx.addIssue({ code: 'custom', path: ['grand_pas'], message: 'le grand pas est plus petit que le pas' });
    }

    if (regle.depart < regle.minimum) {
      ctx.addIssue({ code: 'custom', path: ['depart'], message: 'le depart est sous le minimum' });
    }
  });

// 🔴 Une reponse est toujours fermee — un choix ou un compteur, jamais une saisie de texte.
const question = z
  .object(
    {
      type: z.literal('question'),
      id: identifiant(' — un id relie un item a sa reponse'),
      enonce: texteNonVide(" — le texte de l'item, recopie du corpus, jamais de memoire"),
      precision: texteNonVide().optional(),
      choix: z.array(choix).min(2, { error: 'moins de deux choix — une question est toujours un choix ferme' }).optional(),
      compteur: compteur.optional(),
      reprise: z.boolean({ error: "reprise n'est pas un booleen" }).optional(),
    },
    { error: "ce n'est pas un objet" },
  )
  .superRefine((etape, ctx) => {
    const saisies = [etape.choix, etape.compteur].filter((saisie) => saisie !== undefined).length;

    if (saisies !== 1) {
      ctx.addIssue({
        code: 'custom',
        path: ['choix'],
        message: 'une question porte « choix » OU « compteur », jamais les deux ni aucun des deux',
      });
    }

    if (etape.reprise && !etape.compteur) {
      ctx.addIssue({
        code: 'custom',
        path: ['reprise'],
        message: 'seule une question au compteur se reprend — un choix ferme repart toujours de sa premiere option',
      });
    }
  });

// ⭐ Le seul endroit du dispositif ou une saisie libre est proposee, et elle peut rester vide (R5).
const note = z.object(
  {
    type: z.literal('note'),
    id: identifiant(' — un id relie une note a sa reponse'),
    enonce: texteNonVide(),
    precision: texteNonVide().optional(),
  },
  { error: "ce n'est pas un objet" },
);

const minuteur = z.object(
  {
    type: z.literal('minuteur'),
    secondes: z
      .int({ error: '« secondes » absent ou non entier' })
      .positive({ error: '« secondes » doit etre positif' }),
    consigne: texteNonVide().optional(),
    pour: porteur.optional(),
  },
  { error: "ce n'est pas un objet" },
);

const checklist = z.object(
  {
    type: z.literal('checklist'),
    enonce: texteNonVide(),
    lignes: z.array(texteNonVide(), { error: 'lignes absentes' }).min(1, { error: 'aucune ligne a cocher' }),
  },
  { error: "ce n'est pas un objet" },
);

const confirmation = z.object(
  {
    type: z.literal('confirmation'),
    libelle: texteNonVide(),
  },
  { error: "ce n'est pas un objet" },
);

const etape = z.discriminatedUnion('type', [info, question, note, minuteur, checklist, confirmation], {
  error: `type d'etape inconnu — attendu : ${ETAPES.join(', ')}`,
});

const sortieLibre = z.literal(true, {
  error: 'sortie_libre doit valoir true — sortir avant la fin est toujours permis, sans avoir a le justifier',
});

const relireIdentifiants = (
  etapes: ReadonlyArray<z.infer<typeof etape>>,
  ctx: z.RefinementCtx,
): void => {
  const identifiants = etapes.flatMap((une) => (une.type === 'question' || une.type === 'note' ? [une.id] : []));

  identifiants
    .filter((id, rang) => identifiants.indexOf(id) !== rang)
    .forEach((id) =>
      ctx.addIssue({
        code: 'custom',
        path: ['etapes'],
        message: `etape en double : ${id} — un id relie un item a sa reponse`,
      }),
    );
};

// L'aidant ne peut ni corriger ni improviser : ce qui manque ne s'invente pas, la carte entiere tombe.
const relireCarteTenueParLAidant = (
  carte: { readonly signal_arret?: string; readonly arret?: ReadonlyArray<string>; readonly etapes: ReadonlyArray<z.infer<typeof etape>> },
  ctx: z.RefinementCtx,
): void => {
  const arret = carte.arret || [];
  const dernier = arret[arret.length - 1];
  const minuteurs = carte.etapes.filter((une) => une.type === 'minuteur');

  if (!carte.signal_arret) {
    ctx.addIssue({
      code: 'custom',
      path: ['signal_arret'],
      message:
        "signal_arret absent — Xavier doit pouvoir arreter SANS PARLER, c'est exactement ce qui tombe en premier ; le geste convenu est le « non » de la main, et il se recopie tel quel",
    });
  }

  if (arret.length < 2) {
    ctx.addIssue({
      code: 'custom',
      path: ['arret'],
      message: "au moins deux criteres d'arret, portes par la case a cocher qui ouvre le deroule",
    });
  }

  if (!dernier || !DERNIER_CRITERE.test(dernier)) {
    ctx.addIssue({
      code: 'custom',
      path: ['arret'],
      message:
        "le dernier critere d'arret est toujours « tu ne sais pas quoi faire, on s'arrete » — l'aidant n'improvise jamais",
    });
  }

  if (carte.etapes[0]?.type !== 'checklist') {
    ctx.addIssue({
      code: 'custom',
      path: ['etapes'],
      message:
        "la premiere etape d'une carte tenue par l'aidant est la checklist d'avant — cocher est la seule preuve que les criteres d'arret ont ete lus",
    });
  }

  if (!minuteurs.length) {
    ctx.addIssue({ code: 'custom', path: ['etapes'], message: 'aucune consigne chronometree — le deroule est vide' });
  }

  minuteurs
    .filter((une) => une.pour === undefined || une.consigne === undefined)
    .forEach(() =>
      ctx.addIssue({
        code: 'custom',
        path: ['etapes'],
        message:
          "une consigne tenue par l'aidant porte toujours « pour » et « consigne » — la consigne dit elle-meme a qui elle s'adresse",
      }),
    );
};

const relireCarteTenueParXavier = (
  carte: { readonly signal_arret?: string; readonly arret?: ReadonlyArray<string>; readonly etapes: ReadonlyArray<z.infer<typeof etape>> },
  ctx: z.RefinementCtx,
): void => {
  if (carte.signal_arret !== undefined || carte.arret !== undefined) {
    ctx.addIssue({
      code: 'custom',
      path: ['signal_arret'],
      message: "signal_arret et arret n'ont de sens que sur une carte tenue par l'aidant (porteur: aidant)",
    });
  }

  carte.etapes
    .filter((une) => une.type === 'minuteur' && une.pour !== undefined)
    .forEach(() =>
      ctx.addIssue({
        code: 'custom',
        path: ['etapes'],
        message: "« pour » n'a de sens que sur une carte tenue par l'aidant — Xavier tient la sienne lui-meme",
      }),
    );
};

const panneau = z
  .object({
    ...commun,
    type: z.literal('panneau'),
    rubrique,
    quand,
    porteur: porteur.optional(),
    signal_arret: texteNonVide().optional(),
    arret: z.array(texteNonVide()).optional(),
    sortie_libre: sortieLibre,
    etapes: z
      .array(etape, { error: 'etapes absentes' })
      .min(1, { error: 'aucune etape — une carte tombe entiere ou pas du tout' }),
  })
  .superRefine((carte, ctx) => {
    relireIdentifiants(carte.etapes, ctx);

    if (carte.porteur === 'aidant') {
      relireCarteTenueParLAidant(carte, ctx);
    } else {
      relireCarteTenueParXavier(carte, ctx);
    }

    if (PHQ9.test(carte.id) || PHQ9.test(carte.titre)) {
      ctx.addIssue({
        code: 'custom',
        path: ['titre'],
        message:
          "le PHQ-9 ne se publie jamais — son item 9 interroge l'ideation suicidaire et Kokoro s'interdit tout numero d'urgence par construction ; il se passe en conversation, avec psy-bilan",
      });
    }
  });

// 🔴 Un bilan ne passe pas par la bibliotheque : canal distinct au depot, dossier distinct dans le transit.
// C'est la rubrique bilan qui les separe, et elle va toujours avec une date et sans « quand ».
const pdf = z
  .object({
    ...commun,
    type: z.literal('pdf'),
    rubrique: z.enum(['crise', 'therapie', 'documentation', 'bilan'], {
      error: 'rubrique attendue : crise, therapie, documentation ou bilan',
    }),
    quand: quand.optional(),
    date: jour(' — celle du bilan, jamais celle de la publication').optional(),
    document: identifiant(' — le nom nu du fichier, sans chemin ni extension'),
  })
  .superRefine((carte, ctx) => {
    const bilan = carte.rubrique === 'bilan';

    if (bilan !== (carte.date !== undefined)) {
      ctx.addIssue({
        code: 'custom',
        path: ['date'],
        message: bilan
          ? "un bilan porte toujours sa date — l'ecran Bilan groupe par mois decroissant"
          : "« date » est reservee aux bilans — elle rangerait cette carte sur un ecran qui ne l'affiche pas",
      });
    }

    if (bilan !== (carte.quand === undefined)) {
      ctx.addIssue({
        code: 'custom',
        path: ['quand'],
        message: bilan
          ? "« quand » sur un bilan — sa date appartient au document, pas a l'assiduite de Xavier"
          : '« quand » absent — aujourdhui, au_besoin ou sans_date',
      });
    }
  });

export const CarteSchema = z.discriminatedUnion('type', [panneau, pdf], {
  error: `type inconnu — attendu : ${TYPES.join(', ')}`,
});

export const ProgrammeSchema = z
  .object(
    {
      version: z.int({ error: 'version absente ou non entiere' }),
      publie_le: jour(),
      supervision: texteNonVide(
        ' — rien ne se publie sans une passe du superviseur (superviseur/README.md §4)',
      ),
      cartes: z.array(CarteSchema, { error: 'cartes absentes' }),
    },
    { error: "la racine n'est pas un objet" },
  )
  .superRefine((programme, ctx) => {
    const identifiants = programme.cartes.map((carte) => carte.id);

    identifiants
      .filter((id, rang) => identifiants.indexOf(id) !== rang)
      .forEach((id) =>
        ctx.addIssue({
          code: 'custom',
          path: ['cartes'],
          message: `id en double : ${id} — un id relie une reponse a sa carte, il doit etre unique`,
        }),
      );
  });

export type Carte = z.infer<typeof CarteSchema>;
export type Programme = z.infer<typeof ProgrammeSchema>;
